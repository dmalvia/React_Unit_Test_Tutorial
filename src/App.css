.App {
  font-family: sans-serif;
  text-align: center;
}

.flex {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
}

.flex > div {
  margin: 0 1rem 2rem 1rem;
  text-align: center;
}

img {
  display: inline-block;
  max-width: 100%;
}
.input-box {
  width: 40%;
  padding: 10px;
  font-size: 20px;
  margin: 20px 0px;
}
