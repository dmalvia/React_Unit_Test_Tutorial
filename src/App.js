import React, { useState, useEffect } from "react";
import User from "./User";

const App = () => {
  const [users, setUsers] = useState([]);
  const [searchTerm, setSearchTerm] = useState("");

  const handleChange = ({ target: { value } }) => {
    setSearchTerm(value);
  };

  const fetchUsers = async () => {
    const res = await fetch("https://reqres.in/api/users/", { method: "GET" });
    const json = await res.json();
    setUsers(json.data);
  };

  useEffect(() => {
    fetchUsers();
  }, []);

  return (
    <div className="App">
      <h1>Users</h1>
      <input
        className="input-box"
        type="text"
        placeholder="Search users"
        value={searchTerm}
        onChange={(e) => handleChange(e)}
      />
      <User users={users} searchTerm={searchTerm} />
    </div>
  );
};

export default App;
