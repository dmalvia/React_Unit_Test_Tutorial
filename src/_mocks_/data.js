export const usersData = [
  {
    id: "1",
    first_name: "Dipesh",
    email: "dipesh@gmail.com",
    avatar: "dipesh.jpg",
  },
  {
    id: "2",
    first_name: "Nikesh",
    email: "nikesh@gmail.com",
    avatar: "nikesh.jpg",
  },
];

export const badResponse = {
  type: "error",
  title: "Something went wrong",
};
