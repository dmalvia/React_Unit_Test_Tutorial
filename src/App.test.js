import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { usersData, badResponse } from "./_mocks_/data";
import App from "./App";

const searchValue = "test";
const server = setupServer(
  rest.get("https://reqres.in/api/users/", (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(usersData));
  }),
  rest.get("https://reqres.in/api/users/Bad", (req, res, ctx) => {
    return res(ctx.status(400), ctx.json(badResponse));
  })
);

describe("Test App component", () => {
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it("renders header text as Users", async () => {
    const { container } = render(<App />);
    const headerText = screen.getByText(/Users/i);
    expect(container).toMatchSnapshot();
    expect(headerText).toBeTruthy();
  });

  it("Search bar - input is empty", async () => {
    const { container } = render(<App />);
    const inputSearch = await waitFor(() =>
      screen.findByPlaceholderText(/Search users/i)
    );
    expect(container).toMatchSnapshot();
    expect(inputSearch).toHaveAttribute("value", "");
  });

  it("Search bar - input has value", async () => {
    const { container } = render(<App />);
    const inputSearch = await waitFor(() =>
      screen.findByPlaceholderText(/Search users/i)
    );
    fireEvent.change(inputSearch, { target: { value: searchValue } });
    expect(container).toMatchSnapshot();
    expect(inputSearch).toHaveAttribute("value", searchValue);
  });

  it("return JSON when the response is ok", async () => {
    expect.assertions(3);
    const res = await fetch("https://reqres.in/api/users/", { method: "GET" });
    const data = await res.json();
    expect(res.status).toBe(200);
    expect(res.ok).toBe(true);
    expect(JSON.stringify(data)).toBe(JSON.stringify(usersData));
  });

  it("return JSON when the response is Bad", async () => {
    expect.assertions(3);
    const res = await fetch("https://reqres.in/api/users/Bad", {
      method: "GET",
    });
    const data = await res.json();
    expect(res.status).toBe(400);
    expect(res.ok).toBe(false);
    expect(JSON.stringify(data)).toBe(JSON.stringify(badResponse));
  });
});
