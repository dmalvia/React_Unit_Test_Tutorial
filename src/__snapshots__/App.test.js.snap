// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test App component Search bar - input has value 1`] = `
<div>
  <div
    class="App"
  >
    <h1>
      Users
    </h1>
    <input
      class="input-box"
      placeholder="Search users"
      type="text"
      value="test"
    />
    <div
      class="flex"
    />
  </div>
</div>
`;

exports[`Test App component Search bar - input is empty 1`] = `
<div>
  <div
    class="App"
  >
    <h1>
      Users
    </h1>
    <input
      class="input-box"
      placeholder="Search users"
      type="text"
      value=""
    />
    <div
      class="flex"
    />
  </div>
</div>
`;

exports[`Test App component renders header text as Users 1`] = `
<div>
  <div
    class="App"
  >
    <h1>
      Users
    </h1>
    <input
      class="input-box"
      placeholder="Search users"
      type="text"
      value=""
    />
    <div
      class="flex"
    />
  </div>
</div>
`;
